storage/redis/peer_store.go

```diff
@@ -0,0 +1,773 @@
+
+func (ps *peerStore) ScrapeSwarm(ih bittorrent.InfoHash, af bittorrent.AddressFamily) (resp bittorrent.Scrape) {
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped redis store")
+	default:
+	}
+
+	resp.InfoHash = ih
+	addressFamily := af.String()
+	encodedInfoHash := ih.String()
+	encodedLeecherInfoHash := ps.leecherInfohashKey(addressFamily, encodedInfoHash)
+	encodedSeederInfoHash := ps.seederInfohashKey(addressFamily, encodedInfoHash)
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	leechersLen, err := conn.Do("HLEN", encodedLeecherInfoHash)
+	if err != nil {
+		log.Error("storage: Redis HLEN failure", log.Fields{
+			"Hkey":  encodedLeecherInfoHash,
+			"error": err,
+		})
+		return
+	}
+
+	seedersLen, err := conn.Do("HLEN", encodedSeederInfoHash)
+	if err != nil {
+		log.Error("storage: Redis HLEN failure", log.Fields{
+			"Hkey":  encodedSeederInfoHash,
+			"error": err,
+		})
+		return
+	}
+
+	resp.Incomplete = uint32(leechersLen.(int64))
+	resp.Complete = uint32(seedersLen.(int64))
+
+	return
+}
+
+// collectGarbage deletes all Peers from the PeerStore which are older than the
+// cutoff time.
+//
+// This function must be able to execute while other methods on this interface
+// are being executed in parallel.
+func (ps *peerStore) collectGarbage(cutoff time.Time) error {
+	select {
+	case <-ps.closed:
+		return nil
+	default:
+	}
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	cutoffUnix := cutoff.UnixNano()
+	start := time.Now()
+
+	for _, group := range ps.groups() {
+		// list all infohashes in the group
+		infohashesList, err := conn.Do("HKEYS", group)
+		if err != nil {
+			return err
+		}
+		infohashes := infohashesList.([]interface{})
+
+		for _, ih := range infohashes {
+			ihStr := string(ih.([]byte))
+			isSeeder := len(ihStr) > 5 && ihStr[5:6] == "S"
+
+			// list all (peer, timeout) pairs for the ih
+			ihList, err := conn.Do("HGETALL", ihStr)
+			if err != nil {
+				return err
+			}
+			conIhList := ihList.([]interface{})
+
+			var pk serializedPeer
+			var removedPeerCount int64
+			for index, ihField := range conIhList {
+				if index%2 == 1 { // value
+					mtime, err := strconv.ParseInt(string(ihField.([]byte)), 10, 64)
+					if err != nil {
+						return err
+					}
+					if mtime <= cutoffUnix {
+						ret, err := redis.Int64(conn.Do("HDEL", ihStr, pk))
+						if err != nil {
+							return err
+						}
+
+						removedPeerCount += ret
+
+						log.Debug("storage: deleting peer", log.Fields{
+							"Peer": decodePeerKey(pk).String(),
+						})
+					}
+				} else { // key
+					pk = serializedPeer(ihField.([]byte))
+				}
+			}
+			// DECR seeder/leecher counter
+			decrCounter := ps.leecherCountKey(group)
+			if isSeeder {
+				decrCounter = ps.seederCountKey(group)
+			}
+			if _, err := conn.Do("DECRBY", decrCounter, removedPeerCount); err != nil {
+				return err
+			}
+
+			ihLen, err := conn.Do("HLEN", ihStr)
+			if err != nil {
+				return err
+			}
+			if ihLen.(int64) == 0 {
```

Comment:
This is racy - if someone added a new peer to the infohash `ihStr` _after_ we performed `HLEN` but _before_ we execute `DEL`, we will delete this valid peer. Can we make this atomic?

Also I think we need to decrement the infohash counter if we do delete it.

Reply:
@mrd0ll4r I found one way https://redis.io/topics/transactions and fixed it.

Reply:
Nice! That's pretty much exactly what I had in mind. Didn't know about the WATCH feature, that's pretty nice.

Do we have to wrap the WATCH ... MULTI ... EXEC in a loop in case EXEC fails, i.e. the WATCHed key was modified?

Reply:
Yes, we have to, but we can ignore `EXEC` error as it isn't that critical.

Reply:
I think you're right, but the reason why is quite complex. I'll write this down here, for future reference:

- The `Delete(Seeder|Leecher)` and `GraduateLeecher` methods never delete an infohash _key_ from an addressFamily _hash_. They also never decrement the infohash counter.
- The `Put(Seeder|Leecher)` and `GraduateLeecher` methods only ever add infohash keys to addressFamily hashes and increment the infohash counter.
- The only method that deletes from the addressFamily hashes is `collectGarbage`, which also decrements the counters. That means that, even if a `Delete(Seeder|Leecher)` call removes the last peer from a swarm, the infohash counter is not changed and the infohash is left in the addressFamily hash until it will be cleaned up by `collectGarbage`.
- `collectGarbage` must run regularly.
- A `WATCH ... MULTI ... EXEC` block fails, if between the `WATCH` and the 'EXEC` any of the watched keys have changed. The location of the `MULTI` doesn't matter.

We have to analyze four cases to prove our algorithm works. I'll characterize them by a tuple (`number of peers in a swarm before WATCH`, `number of peers in the swarm during the transaction`).
1. (0,0), the easy case: The swarm is empty, we watch the key, we execute `HLEN` and find it empty. We remove it and decrement the counter. It stays empty the entire time, the transaction goes through. :heavy_check_mark: 
2. (1,n > 0): The swarm is not empty, we watch the key, we find it non-empty, we unwatch the key. All good. No transaction is made, no transaction fails. :heavy_check_mark: 
3. (0,1): We have to analyze this in two ways.
    - If the change happens before the `HLEN` call, we will see that the swarm is not empty and start no transaction. :heavy_check_mark: 
    - If the change happens after the `HLEN`, we will attempt a transaction and it will fail. This is okay, the swarm is not empty, we will try cleaning it up next time `collectGarbage` runs. :heavy_check_mark: 
4. (1,0): Again, two ways:
    - If the change happens before the `HLEN`, we will see an empty swarm. This situation happens if a call to `Delete(Seeder|Leecher)` removed the last peer asynchronously. We will attempt a transaction, but the transaction will fail. This is okay, the infohash _key_ will remain in the addressFamily _hash_, we will attempt to clean it up the next time 'collectGarbage` runs. :heavy_check_mark: 
    - If the change happens after the `HLEN`, we will not even attempt to make the transaction. The infohash _key_  will remain in the addressFamil _hash_ and we'll attempt to clean it up the next time `collectGarbage` runs. :heavy_check_mark: 

This really took me a while to figure out, and it's not trivial to prove. We should maybe note this down somewhere in the code.

Reply:
It's great. How about putting this as `coolectGarbage` comments?

Reply:
Yes, that sounds good, thank you :)